### PR TITLE
crimson:  fix RefCountedObject::_get() assert

### PR DESCRIPTION
**Pull request by liu-chunmei**
fix the bug : https://tracker.ceph.com/issues/61502

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
The commit message isn't sufficient.  What this change actually does is change crimson to use an atomic counter for RefCountedObject instances.  Additionally, always add the redmine bug to the commit message.  Let's change the commit message to:

```
crimson: use atomic counter for RefCountedObject as with classic

crimson-osd uses RefCountedObject instances for objects that move between
reactors, so we really do need to use an atomic counter.

Fixes: https://tracker.ceph.com/issues/61502
Signed-off-by: chunmei <chunmei.liu@intel.com>
```

Btw, which object actually caused the problem?

**Review by athanatos — CHANGES_REQUESTED** [review]
See above

**Comment by liu-chunmei** [comment]
> The commit message isn't sufficient. What this change actually does is change crimson to use an atomic counter for RefCountedObject instances. Additionally, always add the redmine bug to the commit message. Let's change the commit message to:
> 
> ```
> crimson: use atomic counter for RefCountedObject as with classic
> 
> crimson-osd uses RefCountedObject instances for objects that move between
> reactors, so we really do need to use an atomic counter.
> 
> Fixes: https://tracker.ceph.com/issues/61502
> Signed-off-by: chunmei <chunmei.liu@intel.com>
> ```
> 
> Btw, which object actually caused the problem? Presumably the MOSDRepOp embedded in the RepRequest?

I found in two places. one is in RepRequest::with_pg,  call pg->handle_rep_op(req); req refcounter error, one is in primary core  ms_dispatch to send msg to other cores which I added to simulate multicore messenger.

**Comment by liu-chunmei** [comment]
> The commit message isn't sufficient. What this change actually does is change crimson to use an atomic counter for RefCountedObject instances. Additionally, always add the redmine bug to the commit message. Let's change the commit message to:
> 
> ```
> crimson: use atomic counter for RefCountedObject as with classic
> 
> crimson-osd uses RefCountedObject instances for objects that move between
> reactors, so we really do need to use an atomic counter.
> 
> Fixes: https://tracker.ceph.com/issues/61502
> Signed-off-by: chunmei <chunmei.liu@intel.com>
> ```
> 
> Btw, which object actually caused the problem? Presumably the MOSDRepOp embedded in the RepRequest?

done.

**Comment by liu-chunmei** [comment]
@athanatos could you please review it again.

**Comment by athanatos** [comment]
The first line of the commit appears to be  ```crimson/common fix RefCountedObject::_get() assert```.  Remove that so that the short commit description will be ```crimson: use atomic counter for RefCountedObject as with classic```

**Review by athanatos — CHANGES_REQUESTED** [review]
see above

**Review by athanatos — APPROVED** [review]
LGTM, it's a change in common/, so it needs a normal rados suite run, marking needs-qa

**Comment by liu-chunmei** [comment]
> LGTM, it's a change in common/, so it needs a normal rados suite run, marking needs-qa

does the normal rados suite run mean classic ceph ci-testing or crimson ci-testing? this PR doesn't change anything for classic ceph.

**Review by cyx1231st — APPROVED** [review]
LGTM.

> does the normal rados suite run mean classic ceph ci-testing or crimson ci-testing? this PR doesn't change anything for classic ceph.

I think it means the crimson rados test suite.

**Comment by athanatos** [comment]
This is a change to common/, so it needs a full rados suite run, not just the crimson one.

**Comment by liu-chunmei** [comment]
> This is a change to common/, so it needs a full rados suite run, not just the crimson one.

@athanatos  
we use the following command to run  rados suite test,
teuthology-suite -m smithi -e chunmei.liu@intel.com -p 71 -s rados -c wip-chunmei-ceph  
but got 1418214 jobs ,  do you think the command we used  is right?
2023-06-09 01:51:26,864.864 INFO:teuthology.suite.run:Suite rados in /home/yingxin/src/git.ceph.com_ceph-c_wip-chunmei-ceph/qa/suites/rados generated 1418214 jobs (not yet filtered or merged) there 

BTW, the PR not modify ceph branch just affect crimson branch, do we need run 1418214 jobs ?

**Comment by liu-chunmei** [comment]
@athanatos I run crimson-rados suite test, 
the main branch:  https://pulpito.ceph.com/compare/?branch=wip-chunmei-ceph&suite=crimson-rados
this pr branch : https://pulpito.ceph.com/compare/?branch=wip-chunmei-ceph-pr&suite=crimson-rados

each run 2 times, but both branches get different failure in 2 times.  I found it is because osd.x is failed randomly during test in both branches.  so can't tell if this PR introduce new errors. maybe the main branch has other bugs to cause osd.x failed during test. should be fix that bug first.

**Comment by Matan-B** [comment]
Hey @liu-chunmei,

> BTW, the PR not modify ceph branch just affect crimson branch, do we need run 1418214 jobs ?

We don't need to run 1M+ jobs, you can use the `--subset` to have sufficient number of jobs to run.
See: https://docs.ceph.com/projects/teuthology/en/latest/commands/teuthology-suite.html

While the changes affect only Crimson, the file is part of `src/common` so it needs to go through the classic rados suite as well. We skip the classic rados suite testing when the changes are only affect `src/crimson/`.

---

>each run 2 times, but both branches get different failure in 2 times. I found it is because osd.x is failed randomly during test in both branches. so can't tell if this PR introduce new errors. maybe the main branch has other bugs to cause osd.x failed during test. should be fix that bug first.

To find out why the osd is failing you can go to the osd log [1] and find the traceback.
You can check if the failure is a known issue or a new one if it is listed in the tracker.
https://tracker.ceph.com/projects/crimson/issues

The most common failures we are experiencing atm and most likely to cause the crash are:
* https://tracker.ceph.com/issues/61504
* https://tracker.ceph.com/issues/59165
Which are WIP and should be merged soon and result in a better testing coverage.

[1] osd log dir example from last run: http://qa-proxy.ceph.com/teuthology/matan-2023-06-12_07:24:17-crimson-rados-wip-matanb-crimson-testing-10-june-distro-crimson-smithi/7301304/remote/smithi002/log/

**Comment by liu-chunmei** [comment]
@Matan-B Thanks Matan, BTW , in command   --subset <index/outof> what the index/outof should be? can you give me the whole command example? I never run it before.

**Comment by Matan-B** [comment]
> @Matan-B Thanks Matan, BTW , in command --subset <index/outof> what the index/outof should be? can you give me the whole command example? I never run it before.

Sure, for instance:
`teuthology-suite -m smithi -s rados -p <> --sha1 <> --subset 1/150000 `
It's also possible to limit the number of jobs run with `-l <>`.

File: src/common/RefCountedObj.h [hunk 1]
```diff
@@ -74,12 +74,7 @@ class RefCountedObject {
 private:
   void _get() const;
 
-#if defined(WITH_SEASTAR) && !defined(WITH_ALIEN)
-  // crimson is single threaded at the moment
-  mutable uint64_t nref{1};
-#else
   mutable std::atomic<uint64_t> nref{1};
```
Comment (rzarzynski) [inline src/common/RefCountedObj.h:77]: How broad the change is? From the comments I understand there are 2 users hit by the lack of the atomic. I'm curious how much we're sacrificing to rectify them.

**Comment by rzarzynski** [comment]
Adding the `peformance` label as the change might have an impact.

File: src/common/RefCountedObj.h (hunk 1 above)
Comment (liu-chunmei) [reply src/common/RefCountedObj.h:77]: it maybe happened when you transfer conn or msg from one shard to another shard by value copy.

**Comment by liu-chunmei** [comment]
Ceph ci test : 
without this PR: https://pulpito.ceph.com/yingxin-2023-06-14_08:35:09-rados-wip-chunmei-ceph-distro-default-smithi/
with this PR:  https://pulpito.ceph.com/yingxin-2023-06-14_08:38:48-rados-wip-chunmei-ceph-pr-distro-default-smithi/ 
seems not introduce new error. 
@athanatos @Matan-B

**Pull request merged**